### ISSUE TITLE
Issue with the validation_data with unsupervised and semi-supervised learning using generators

### ISSUE THREAD

**Issue by madtoinou** [voxelmorph/voxelmorph#248]
**Task**

I am trying to obtain the validation loss of my model during the training (call to fit). We work with worms brain data, some of the volume are segmented.

**What have you tried**

For unsupervised, it works when we use arrays as input ("generator" similar to the one presented in the tutorial) and arrays as validation_data (homemade function, inspired by the generated mentioned above). It works but it's very slow and limited by the memory.

For semi-supervised and "unsupervised 2.0", we use the generator found in your repository (respectively semisupervised and volgen) for the x data but when we use the same generator for the validation_data, at the end of the first epoch, the fitting is stuck in an infinite loop. The same occurs if we use the "arrays generator" used with unsupervised 1.0.

The training work just fine when the argument validation_data is left empty.

**Details of experiments**

We want to train semi-supervised with a mix of segmented (segmentation with integer values between 0 and 7) and non-segmented data (segmentation is array of zeros)

**Comment by adalca** [comment]
@madtoinou  sorry for missing this earlier

> For unsupervised, it works when we use arrays as input ("generator" similar to the one presented in the tutorial) and arrays as validation_data (homemade function, inspired by the generated mentioned above). It works but it's very slow and limited by the memory.

is it slow without the validation generator?

> For semi-supervised and "unsupervised 2.0", we use the generator found in your repository (respectively semisupervised and volgen) for the x data but when we use the same generator for the validation_data, at the end of the first epoch, the fitting is stuck in an infinite loop or return val_loss equals to 0 at the end of each epoch. This also occurs when the semisupervised generator is called in a comprehension list in order to "convert it".

I'm not sure I fully understand, would you mind providing some code? @ahoopes  did you happen to see any of this?

**Comment by ahoopes** [comment]
Yes, if we could see some example code it would help in replicating the problem. So, it seems like every thing works okay until you specify validation data? In which case it runs forever? Are you using a generator-style data loader without specifying the `validation_steps` parameter? The generators that are provided in voxelmorph will in theory run forever if there is no stopping measure
